bigquery/gcloud/aio/bigquery/utils.py

```diff
@@ -0,0 +1,119 @@
+import datetime
+import logging
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import List
+
+
+log = logging.getLogger(__name__)
+
+
+def flatten(x: Any) -> Any:
+    """
+    Flatten response objects into something we can actually work with.
+
+    The API returns data of the form:
+
+        {'f': [{'v': ...}]}
+
+    to indicate groupings of fields and their potential for having multiple
+    values. We want those to just be plain old objects.
+    """
+    if isinstance(x, dict):
+        # TODO: what if a user has stored a dictionary in their table and that
+        # dictionary is shaped the same way as Google's response format?
+        if 'f' in x:
+            return [flatten(y['v']) for y in x['f']]
+
+        if 'v' in x:
+            return flatten(x['v'])
+
+    if isinstance(x, list):
+        return [flatten(y) for y in x]
+
+    return x
+
+
+def parse(field: Dict[str, Any], value: Any) -> Any:
+    """
+    Parse a given field back to a Python object.
+
+    This is often trivial: convert the value from a string to the type
+    specified in the field's schema. There's a couple caveats we've identified
+    so far, though:
+
+    * NULLABLE fields should be handled specially, eg. so as not to
+      accidentally convert them to the schema type.
+    * REPEATED fields are nested a biot differently than expected, so we need
+      to flatten *first*, then convert.
+
+    `Field = Dict[str, Union[str, 'Field']]`, but wow is that difficult to
+    represent in a backwards-enough compatible fashion.
+    """
+    try:
+        f: Callable[[Any], Any] = {  # type: ignore[assignment]
```

Comment:
Can we call this variable something else? Makes it hard to comprehend as there is `f` being used in the list/dict unpacking as well.

Reply:
Yeah, good call.